pandas/core/frame.py

```diff
@@ -2216,10 +2216,26 @@ def to_feather(self, path, **kwargs) -> None:
         """,
     )
     def to_markdown(
-        self, buf: Optional[IO[str]] = None, mode: Optional[str] = None, **kwargs
+        self,
+        buf: Optional[IO[str]] = None,
+        mode: Optional[str] = None,
+        index: Optional[bool] = True,
+        **kwargs,
     ) -> Optional[str]:
+        if "showindex" in kwargs:
+            index = kwargs["showindex"]
+            del kwargs["showindex"]
```

Comment:
In previous discussion, @jreback  and @TomAugspurger have said that the expected behaviour in the future is that ```showindex``` would be an invalid kwarg. These lines are just a help for the users while we are just raising warnings. So maybe in a future version we would remove these two lines and only raise a ```ValueError``` instead.
```kwargs.setdefault("showindex",...)``` is just for composing kwargs for the underlying tabulate lib.

Reply:
> These lines are just a help for the users while we are just raising warnings.

AFAICT tests pass without this code. How do these lines help users?

Reply:
Isn't ```FutureWarning``` means ```showindex``` is not valid anymore but we'll fix it ourselves for you this time?

Reply:
Isn't ```FutureWarning``` means ```showindex``` is not valid anymore but we'll fix it ourselves for you this time?

Reply:
> So maybe in a future version we would remove these two lines and only raise a `ValueError` instead.

I still don't see the need for these 2 lines of code. At present, they are not needed and in the future will still not be needed since an exception will be raised instead of a warning.

Reply:
At the present, they are the part where we see that the user specify ```showindex``` and divert it to our ```index``` param. In the future it is indeed not needed and the if clause would become only
```py
if "showindex" in kwargs:
    raise....
```

Reply:
Currently, it is
```py
if "showindex" in kwargs:
    index=showindex
    warn...
```

Reply:
> At the present, they are the part where we see that the user specify `showindex` and divert it to our `index` param.

isn't that taken care of by `kwargs.setdefault("showindex", index)` anyway? ie. use showindex if given else use index?

Reply:
Oh, gotcha. I'm just trying to keep all the ```showindex```-related code in one place so in the future when we refactor, the only place we need to care is inside the if clause. But if you want, i could rewrite it as the way you said.

Reply:
> But if you want, i could rewrite it as the way you said.

I don't think anything else needs to change. if you delete the 2 lines, do the tests pass?

Reply:
Yah, the tests pass without the two lines, I overthought and forget that the whole **kwarg is passed down to tabulate